tidy3d/plugins/autograd/invdes/filters.py

```diff
@@ -1,71 +1,226 @@
+    ----------
+    kernel_size : Tuple[pd.PositiveInt, ...]
+        Size of the kernel in pixels for each dimension.
+    normalize : bool = True
+        Whether to normalize the kernel so that it sums to 1.
+    padding : PaddingType = "reflect"
+        The padding mode to use.
+    """
+
+    kernel_size: Union[pd.PositiveInt, Tuple[pd.PositiveInt, ...]] = pd.Field(
+        ..., description="Size of the kernel in pixels for each dimension."
+    )
+    normalize: bool = pd.Field(
+        True, description="Whether to normalize the kernel so that it sums to 1."
+    )
+    padding: PaddingType = pd.Field("reflect", description="The padding mode to use.")
+
+    @classmethod
+    def from_radius_dl(
+        cls, radius: Union[float, Tuple[float, ...]], dl: Union[float, Tuple[float, ...]], **kwargs
+    ) -> "AbstractFilter":
+        """Create a filter from radius and grid spacing.
+
+        Parameters
+        ----------
+        radius : Union[float, Tuple[float, ...]]
+            The radius of the kernel. Can be a scalar or a tuple.
+        dl : Union[float, Tuple[float, ...]]
+            The grid spacing. Can be a scalar or a tuple.
+        **kwargs
+            Additional keyword arguments to pass to the filter constructor.
+
+        Returns
+        -------
+        AbstractFilter
+            An instance of the filter.
+        """
+        kernel_size = get_kernel_size_px(radius=radius, dl=dl)
+        return cls(kernel_size=kernel_size, **kwargs)
+
+    @staticmethod
+    @abc.abstractmethod
+    def get_kernel(size_px: Iterable[int], normalize: bool) -> NDArray:
+        """Get the kernel for the filter.
+
+        Parameters
+        ----------
+        size_px : Iterable[int]
+            Size of the kernel in pixels for each dimension.
+        normalize : bool
+            Whether to normalize the kernel so that it sums to 1.
+
+        Returns
+        -------
+        NDArray
+            The kernel.
+        """
+        ...
+
+    def __call__(self, array: NDArray) -> NDArray:
+        """Apply the filter to an input array.
+
+        Parameters
+        ----------
+        array : NDArray
+            The input array to filter.
+
+        Returns
+        -------
+        NDArray
+            The filtered array.
+        """
+        original_shape = array.shape
+        squeezed_array = np.squeeze(array)
+        size_px = tuple(np.atleast_1d(self.kernel_size))
+        print(size_px)
+        if len(size_px) != squeezed_array.ndim:
+            size_px *= squeezed_array.ndim
+        kernel = self.get_kernel(size_px, self.normalize)
+        convolved_array = convolve(squeezed_array, kernel, padding=self.padding)
+        return np.reshape(convolved_array, original_shape)
+
+
+class ConicFilter(AbstractFilter):
+    """A conic filter for creating and applying convolution filters."""
+
+    @staticmethod
+    @lru_cache(maxsize=1)
+    def get_kernel(size_px: Iterable[int], normalize: bool) -> NDArray:
```

Comment:
can this be a `@cached_property`?

Reply:
No because the kernel dimensionality is decided at runtime from the array that is passed in, so it needs to be a method. And a method with `self` as an argument shouldn't be cached because the cache will keep a reference to the object, meaning it can't be garbage collected, leading to memory leaks. Hence why it is a `staticmethod` (no reference to `self`).